src/main/java/com/googlecode/download/maven/plugin/internal/cache/FileBackedIndex.java

```diff
@@ -94,7 +94,7 @@ private static void create(final File store) {
      * @param store file where index is persisted.
      */
     @SuppressWarnings("unchecked")
-    private void loadFrom(final File store) {
+    private synchronized void loadFrom(final File store) {
```

Comment:
@leonard84 to make FileBackedIndex thread safe we need to declare as `synchronized` **every public method which access (read or write) object state** (`index` or `storage`) fields.

Otherwise, access to those state fields by thread 1 may be not visible to thread 2.
Also 2 operations (such as `this.index.put` and `this.save()`) are not atomic when observed from other thread. Huge source of races :(

Reply:
done